Implement operand bundles for floating-point operations

spavloff opened the pull request:
Currently floating-point operations in general form (beyond the default mode) are always represented by calls to constrained intrinsics. In addition to the side effect, they carry additional information in the form of metadata arguments. This scheme is not efficient in the case of intrinsic function calls, as was noted in
https://discourse.llvm.org/t/thought-on-strictfp-support/71453, because it requires defining a separate intrinsic for the same operation but used in non-default FP environment. The solution proposed in the discussion was "to move the complexity about the environment tracking from the intrinsics themselves to the call instruction".

The way implemented in this change is to use operand bundles (https://llvm.org/docs/LangRef.html#operand-bundles). This way was tried previously (https://reviews.llvm.org/D93455), but was not finished.

This change does not add any new functionality, it only adds the new way of keeping FP related information in LLVM IR. Metadata arguments of constrained functions are preserved, but they are not used in the queries like `getRoundingMode` or `getExceptionBehavior`.

spavloff committed:
Implement operand bundles for floating-point operations

Currently floating-point operations in general form (beyond the default
mode) are always represented by calls to constrained intrinsics. In
addition to the side effect, they carry additional information in the
form of metadata arguments. This scheme is not efficient in the case of
intrinsic function calls, as was noted in
https://discourse.llvm.org/t/thought-on-strictfp-support/71453, because
it requires defining a separate intrinsic for the same operation but
used in non-default FP environment. The solution proposed in the
discussion was "to move the complexity about the environment tracking
from the intrinsics themselves to the call instruction".

The way implemented in this change is to use operand bundles
(https://llvm.org/docs/LangRef.html#operand-bundles). This way was tried
previously (https://reviews.llvm.org/D93455), but was not finished.

This change does not add any new functionality, it only adds the new way
of keeping FP related information in LLVM IR. Metadata arguments of
constrained functions are preserved, but they are not used in the
queries like `getRoundingMode` or `getExceptionBehavior`.

arsenm commented on llvm/lib/IR/AutoUpgrade.cpp:
```diff
@@ -4282,6 +4282,64 @@ static void upgradeDbgIntrinsicToDbgRecord(StringRef Name, CallBase *CI) {
 }
 
+static CallBase *upgradeConstrainedIntrinsicCall(CallBase *CB, Function *F,
+                                                 IRBuilder<> &Builder) {
```
Missing auto upgrade tests

arsenm commented on llvm/lib/IR/AutoUpgrade.cpp:
```diff
@@ -4282,6 +4282,64 @@ static void upgradeDbgIntrinsicToDbgRecord(StringRef Name, CallBase *CI) {
+  CallInst *NewCB = nullptr;
+  if (!NewBundles.empty()) {
+    SmallVector<Value *, 4> Args(CB->args());
+    SmallVector<OperandBundleDef, 2> Bundles;
```
Should be removing the old arguments?

spavloff replied:
It requires change of the called function, as metadata arguments are mandatory. Eventually it should be implemented but in a subsequent patch.

arsenm commented on llvm/lib/IR/AutoUpgrade.cpp:
```diff
@@ -4282,6 +4282,64 @@ static void upgradeDbgIntrinsicToDbgRecord(StringRef Name, CallBase *CI) {
+    if (CurrentEB) {
+      assert(*EB == *CurrentEB);
+    } else {
+      NewBundles.emplace_back("fpe.except",
+                              ConstantInt::get(Type::getInt32Ty(C), *EB));
```
Probably should omit the bundle if it is the default / strictest setting

spavloff replied:
Probably. But if a function is used in strictfp environment, it must have a side effect for proper ordering. Keeping explicit bundles could assist this.

arsenm commented on llvm/lib/IR/AutoUpgrade.cpp:
```diff
@@ -4282,6 +4282,64 @@ static void upgradeDbgIntrinsicToDbgRecord(StringRef Name, CallBase *CI) {
+    } else {
+      int RMValue = static_cast<int>(*RM);
+      NewBundles.emplace_back("fpe.round",
+                              ConstantInt::get(Type::getInt32Ty(C), RMValue));
```
Is it possible to continue encoding these as the same metadata values, instead of the raw integer?

spavloff replied:
This is in progress. Now metadata cannot be used in bundles.

arsenm commented on llvm/lib/IR/AutoUpgrade.cpp:
```diff
@@ -4282,6 +4282,64 @@ static void upgradeDbgIntrinsicToDbgRecord(StringRef Name, CallBase *CI) {
+    Bundles.append(NewBundles);
+
+    Builder.SetInsertPoint(CB->getParent(), CB->getIterator());
+    MDNode *FPMath = CB->getMetadata(LLVMContext::MD_fpmath);
```
There's no point in querying fpmath if you just use copyMetadata below.

I would expect you could also just hack up the call site in place without creating a fresh call instead.

arsenm reviewed (commented):
I think the arguments should be dropped from the intrinsics if they're being moved to bundles

andykaylor commented on llvm/include/llvm/IR/FPEnv.h:
```diff
@@ -43,6 +43,15 @@ enum ExceptionBehavior : uint8_t {
 
 }
 
+inline bool isValidExceptionBehavior(unsigned X) {
```
Do you know why the exception behaviors aren't handled in FloatingPointMode.h?

spavloff replied:
I think this is historically, because the functionality was implemented by different people. We have "FPE.h" and "FloatingPointMode.h", both describing FP environment.

andykaylor commented on llvm/include/llvm/IR/IRBuilder.h:
```diff
@@ -357,6 +357,9 @@ class IRBuilderBase {
 
   void setConstrainedFPCallAttr(CallBase *I) {
     I->addFnAttr(Attribute::StrictFP);
+    MemoryEffects ME = MemoryEffects::inaccessibleMemOnly();
```
I think all of the constrained intrinsics currently use inaccessibleMemOnly, but it might be a good idea to get this attribute from the intrinsic itself. If we added an intrinsic for something like sincos, it would access argument memory too.

We should consider adding separate memory effects for fp control and status registers. Obviously that would be a separate change set.

arsenm replied:
It shouldn't be necessary to touch the attributes. The set of intrinsic attributes are fixed (callsite attributes are another thing, but generally should be droppable here)

andykaylor replied:
From the LangRef "Calls and invokes with operand bundles have unknown read / write effect on the heap on entry and exit (even if the call target specifies a memory attribute), unless they’re overridden with callsite specific attributes." I assumed that's why the callsite attribute is being set here.

spavloff replied:
If a constrained intrinsic like `experimental_constrained_sqrt` is used, its memory effect is taken from the intrinsic properties. If `sqrt` is used with bundle operands, the memory effect must be set explicitly.

andykaylor commented on llvm/docs/LangRef.rst:
```diff
@@ -2994,6 +2994,29 @@ A "convergencectrl" operand bundle is only valid on a ``convergent`` operation.
 
+.. _ob_fpe:
+
+Floating-point Environment Operand Bundles
```
Should we place restrictions on which intrinsics these operand bundles can be attached to? It seems that they would have no meaning for non-fp intrinsics, and it's unclear whether they should require the "strictfp" attribute.

spavloff replied:
I think we could introduce additional "memory" resources in addition to inaccessible memory to represent FP environment. In this case intrinsisc like `sqrt` would have effect like "reads control modes" and "sets status bits", which would characterize FP intrinsics. In the case of default FP environment these properties should be erased however.

Hopefully more fine-grained FP environment access could avoid unneeded ordering and increase performance of non-default execution modes.

andykaylor reviewed (commented):
I don't think attaching these operand bundles to the existing set of intrinsics is what we want to do. I'd prefer to deprecate the existing intrinsics and define some new scheme that uses these bundles on intrinsics that may be used in other ways.

One of the original reasons that we talked about the idea of using operand bundles this way is that we wanted a way to indicate strict FP constraints on vector predication intrinsics without requiring an entire second set of intrinsics. I'd like to see a set of generic floating point intrinsics (possibly the existing non-constrained intrinsics) that could be modified with operand bundles in the way you've implemented here, but I guess we'd need some mechanism to ensure that the operand bundles were respected.

arsenm commented:
Also it's silly that we need to do bitcode autoupgrade of "experimental" intrinsics, but x86 started shipping with strictfp enabled in production before they graduated. We might as well drop the experimental bit then

spavloff committed:
Use metadata for bundle values

kpneal commented:
> I don't think attaching these operand bundles to the existing set of intrinsics is what we want to do. I'd prefer to deprecate the existing intrinsics and define some new scheme that uses these bundles on intrinsics that may be used in other ways.
> 
> One of the original reasons that we talked about the idea of using operand bundles this way is that we wanted a way to indicate strict FP constraints on vector predication intrinsics without requiring an entire second set of intrinsics. I'd like to see a set of generic floating point intrinsics (possibly the existing non-constrained intrinsics) that could be modified with operand bundles in the way you've implemented here, but I guess we'd need some mechanism to ensure that the operand bundles were respected.

I thought bundles were a way to allow us to keep the FP-instruction-constrained intrinsics but dump all the other ones, and the bundles could be attached to any call that needed constrained semantics? I really, really hope we're keeping the safe-by-default semantics of the constrained intrinsics. I can't tell you how important the constrained semantics are to some compiler users like myself, and having to spend the next forever shooting down bugs because the constrained semantics aren't respected in some cases would be seriously damaging.

The IR Verifier changes aren't in place yet in part because there are I don't even remember how many hundreds of places that need to be corrected. A violation of constrained semantics wouldn't be caught by the Verifier (say, if something was hoisted) so we'd never be able to be confident that we got it right and kept it right.

If we can't keep the constrained semantics and near-100% guarantee that no new exceptions will be introduced then operand bundles are not a replacement for the constrained intrinsics.

arsenm commented:
> If we can't keep the constrained semantics and near-100% guarantee that no new exceptions will be introduced then operand bundles are not a replacement for the constrained intrinsics.

We would still need a call / function attribute to indicate strictfp calls, and such calls would then be annotatable with bundles to relax the assumptions. The default would always have to be the most conservative assumption

kpneal commented:
> > If we can't keep the constrained semantics and near-100% guarantee that no new exceptions will be introduced then operand bundles are not a replacement for the constrained intrinsics.
> 
> We would still need a call / function attribute to indicate strictfp calls, and such calls would then be annotatable with bundles to relax the assumptions. The default would always have to be the most conservative assumption

With the constrained intrinsics the default is safe because optimizations don't recognize the constrained intrinsic and thus don't know how to optimize it. If we instead rely on the strictfp attribute then we'll need possibly thousands of checks for this attribute, we'll need everyone going forward to remember to check for it, and we'll have no way to verify that this rule is being followed.

jcranmer-intel commented on llvm/docs/LangRef.rst:
```diff
@@ -2994,6 +2994,29 @@ A "convergencectrl" operand bundle is only valid on a ``convergent`` operation.
+Floating-point Environment Operand Bundles
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+These operand bundles provide details on how the operation interacts with the
+:ref:`floating-point environment <_floatenv>`. There are two kinds of such
+operand bundles, which characterize interaction with floating-point control
+modes and status bits.
```
It seems to me that we also need the ability to represent that the instructions may depend on more than just the dynamic rounding mode for effect, for example denormal flushing bits, exception masking bits, or more exotic target-specific stuff (e.g., x87's precision control bit).

andykaylor replied:
> more exotic target-specific stuff (e.g., x87's precision control bit)

For anyone who doesn't think FTZ complicates their life enough.....

We seem to be currently taking the position that FTZ/DAZ is outside the scope of the IR semantics, though I may have mentioned that I'm not entirely happy with that. An even stronger case could be made for pretending the x87 precision control doesn't exist. We have IR types for binary32, binary64, and x86_fp80. It wouldn't be entirely unreasonable to say that the behavior of the latter is unspecified if the precision control bit is changed. I'm not sure there's a realistic use case for dynamically changing the precision control, and clang already avoids using the x86_fp80 type on Windows, where the default setting isn't 80-bit. I guess someone could have legacy code that does it.

andykaylor commented:
> With the constrained intrinsics the default is safe because optimizations don't recognize the constrained intrinsic and thus don't know how to optimize it. If we instead rely on the strictfp attribute then we'll need possibly thousands of checks for this attribute, we'll need everyone going forward to remember to check for it, and we'll have no way to verify that this rule is being followed.

I may have given a wrong impression here. I'm basically in agreement with you on this point. We definitely need the safe-by-default behavior. What I meant to say was that we don't necessarily need the current constrained intrinsics. We could replace the current constrained intrinsics with a new set of intrinsics that could be used for this and other purposes.

Right now, I can think of at least three types of math intrinsics needed:

1. Correctly-rounded results
2. Vector-predication
3. Strict FP semantics

I think a lot of us would wish that the semantics of the libm-equivalent intrinsics required correctly rounded results, but as long as the common system math libraries aren't producing correctly-rounded results, that's a tough ask.

I'm a bit concerned that we've missed the boat on the vector-predicated intrinsics, because they are in widespread use already, so if we want to combine vector predication with strict FP semantics via operand bundles, that gets us right back into the problem of retrofitting correctness.

spavloff commented on llvm/docs/LangRef.rst:
```diff
@@ -2994,6 +2994,29 @@ A "convergencectrl" operand bundle is only valid on a ``convergent`` operation.
+modes and status bits.
+
+An operand bundle tagged with "fpe.round" may be associated with the operations
+that may depend on rounding mode. It has an integer value, which represents
```
I am working on using metadata for these bundles.

spavloff commented on llvm/include/llvm/ADT/FloatingPointMode.h:
```diff
@@ -47,6 +47,15 @@ enum class RoundingMode : int8_t {
 };
 
+inline bool isValidRoundingMode(int X) {
+  return X >= 0 && X <= static_cast<int>(RoundingMode::Dynamic);
```
This is bad function name. Actually set of supported rounding modes depends on the used target. For example not every target supports rounding `NearestTiesToAway`. But any rounding mode must fit 3 bits. So the function is about loose validation.

If metadata will be used fo "fpe.round", this function would be unnecessary.

arsenm commented:
> With the constrained intrinsics the default is safe because optimizations don't recognize the constrained intrinsic and thus don't know how to optimize it. If we instead rely on the strictfp attribute then we'll need possibly thousands of checks for this attribute, we'll need everyone going forward to remember to check for it, and we'll have no way to verify that this rule is being followed.

The current state already requires you to check this for any library calls. Not sure any wide audit of those ever happened. I don't see a better alternative to cover those, plus the full set of target intrinsics.

kpneal commented:
> It does not look as a good base for safety. Many users want the code in non-default FP modes to be more efficient, Now any any deviation from the default mode requires use of constrained intrinsics in the entire function, but now such solution exhibits poor performance, unsuitable for practical needs. Eventually optimizations will involve the constrained intrinsics too.

This sounds like a rejection of safe-by-default. Keep in mind that "unsuitable for practical needs" doesn't mean that everyone agrees with you about what level of performance is unsuitable for one's practical needs. And a compiler that generates good performing code with random FP exceptions inserted is, in my experience, unusable if one runs with traps on. Trading performance for exception safety is a valid choice for some. Trading away all of the performance lost because few optimization passes know how to deal with the constrained intrinsics is still a valid choice for some.

WRT eliminating the constrained intrinsics completely, I thought that operand bundles could only be attached to function calls and not regular instructions? If I'm wrong, we _still_ have a problem because there are so many uses of the regular FP instructions that we can't be safe-by-default and still use those instructions. We'd need to keep some kind of the constrained intrinsics (or new intrinsics) that give us replacements for the regular FP instructions.

I do like the idea of finding ways to get the behavior we need without having to have most developers care about strictfp. Reality is that most LLVM developers don't have any interest in strictfp. If we're piggybacking off of other parts of the LLVM design that are already widely supported then that makes our jobs much easier. If we can rely on marking calls as "inaccessiblememonly" or other side effects then that sounds quite helpful.

The problem I've run into with the IR Verifier is that the strictfp mode of the IR Builder needs to be turned on in all cases needed, but currently doesn't because the IR Builder's constructor doesn't know to construct itself in strictfp mode. Now, many instantiations of the IR Builder occur with the constructor able to chase down the function definition and find out if strictfp mode is required. Many, hundreds, don't have that luxury and thus each has to be touched and corrected.

Having the IR Builder keep this strictfp mode and automatically add side-effect notes and/or operand bundles when required allows most developers to continue to not care about strictfp but preserve strictfp functionality at the same time. We'd still need a way to find out that some cases were missed because, for example, a strictfp attribute was forgotten. So the IR Builder instantiations still need to be corrected, but that's work in progress. Eventually the IR Builder could be changed to add bundles to, I don't know, all function calls that have a FP type passed to them perhaps?

No IR Builder changes are needed for this patch, of course, but I do want us to make sure we're keeping the whole design in mind as we go.

spavloff commented:
> This sounds like a rejection of safe-by-default. 

What do you mean by "safe-by-default"? Implementation must be correct, and produce the code that executes in accordance to standards and documentation. Any deviation from them is an error and should be fixed. Does it mean that such implementation would be "safe"? 

> Keep in mind that "unsuitable for practical needs" doesn't mean that everyone agrees with you about what level of performance is unsuitable for one's practical needs. And a compiler that generates good performing code with random FP exceptions inserted is, in my experience, unusable if one runs with traps on. 

Users have different needs and can tune compilation using appropriate options. Those who need strict exception behavior chose suitable option set and should get code that behave as expected but probably has lower performance. And there is a large number of users who need maximum performance but can sacrifice exception strictness. They should use another set of options. Does using different compilation options solves this problem?

> Trading performance for exception safety is a valid choice for some. Trading away all of the performance lost because few optimization passes know how to deal with the constrained intrinsics is still a valid choice for some.

There is no hope to produce code as performant as in default mode, but with strict exception semantics. So the only way to make a compiler suitable for all users is to provide means to tune it. 

> WRT eliminating the constrained intrinsics completely, I thought that operand bundles could only be attached to function calls and not regular instructions? If I'm wrong, we still have a problem because there are so many uses of the regular FP instructions that we can't be safe-by-default and still use those instructions. We'd need to keep some kind of the constrained intrinsics (or new intrinsics) that give us replacements for the regular FP instructions.

There is no intention to remove constrained intrinsics (like `experimental.constrained.fadd`), because operand bundles could only be attached to function calls. The constrained intrinsics however need to change their shape in future to use operand bundles as well.

> I do like the idea of finding ways to get the behavior we need without having to have most developers care about strictfp. Reality is that most LLVM developers don't have any interest in strictfp. If we're piggybacking off of other parts of the LLVM design that are already widely supported then that makes our jobs much easier. If we can rely on marking calls as "inaccessiblememonly" or other side effects then that sounds quite helpful.

The new representation does not differ in this aspect from the existing one. FP operations are represented by function calls, they have side effect and proper ordering must be provided by the existing LLVM mechanisms, a developer does not need to know about these properties. There is a hope that using side effect more fine-grained than "inaccessiblememonly" could make FP code faster for non-default modes, but it is in future.

> The problem I've run into with the IR Verifier is that the strictfp mode of the IR Builder needs to be turned on in all cases needed, but currently doesn't because the IR Builder's constructor doesn't know to construct itself in strictfp mode. Now, many instantiations of the IR Builder occur with the constructor able to chase down the function definition and find out if strictfp mode is required. Many, hundreds, don't have that luxury and thus each has to be touched and corrected.

If IRBuilder does not set strictfp when it should, it usually ends up with assertion violation or verification errors. I guess in this case IR is created by some tool, probably MLIR-based, not by clang? Maybe this is a use case that we do not support enough well now and it would be helpful to investigate it.

arsenm commented:
> WRT eliminating the constrained intrinsics completely, I thought that operand bundles could only be attached to function calls and not regular instructions? If I'm wrong, we _still_ have a problem because there are so many uses of the regular FP instructions that we can't be safe-by-default and still use those instructions. We'd need to keep some kind of the constrained intrinsics (or new intrinsics) that give us replacements for the regular FP instructions.

Right, we would need to introduce new llvm.fadd etc. to carry bundles. If there are no bundles these could fold back to the regular instruction

kpneal commented:
> > WRT eliminating the constrained intrinsics completely, I thought that operand bundles could only be attached to function calls and not regular instructions? If I'm wrong, we _still_ have a problem because there are so many uses of the regular FP instructions that we can't be safe-by-default and still use those instructions. We'd need to keep some kind of the constrained intrinsics (or new intrinsics) that give us replacements for the regular FP instructions.
> 
> Right, we would need to introduce new llvm.fadd etc. to carry bundles. If there are no bundles these could fold back to the regular instruction

Good. The new intrinsics will be unknown to the optimization passes and thus be "safe by default". 

The target-specific intrinsics with bundles will be "unsafe by default", but that's a step up from the current just plain unsafe.

We will need someone to go through and audit the libm-style intrinsics and make sure all optimizations of them are safe before we can switch over. Those are "unsafe by default" since the optimizers know about them already.

Folding back to regular instructions will be fairly limited I expect until perhaps very late in the target-specific pipeline. Otherwise we run into the instruction reordering issue. The outliner can probably sometimes extract the non-strictfp parts of a function and mark the new outlined function as not being strictfp which would then enable folding back to instructions.

Do we know how we're going to get bundles through the SelectionDAG?

kpneal commented:
> > This sounds like a rejection of safe-by-default.
> 
> What do you mean by "safe-by-default"? Implementation must be correct, and produce the code that executes in accordance to standards and documentation. Any deviation from them is an error and should be fixed. Does it mean that such implementation would be "safe"?

When I say "safe-by-default" I'm referring to the case where the optimizers don't know about the intrinsics so the optimizers get very conservative and don't optimize. 

It is assumed that all optimizations that actually are done will be in accordance with standards and documentation. 

> > Keep in mind that "unsuitable for practical needs" doesn't mean that everyone agrees with you about what level of performance is unsuitable for one's practical needs. And a compiler that generates good performing code with random FP exceptions inserted is, in my experience, unusable if one runs with traps on.
> 
> Users have different needs and can tune compilation using appropriate options. Those who need strict exception behavior chose suitable option set and should get code that behave as expected but probably has lower performance. And there is a large number of users who need maximum performance but can sacrifice exception strictness. They should use another set of options. Does using different compilation options solves this problem?

Aren't you describing the current state of clang/llvm? Users have different compiler options to choose between FP trap safe and trap-unsafe. Compiled code that is FP trap safe has lower performance. 

The sticky point is the rounding mode. Running with a different rounding mode ideally wouldn't affect performance much at all, but it currently does. But fixing that requires fixing all the optimization passes. 

> > Trading performance for exception safety is a valid choice for some. Trading away all of the performance lost because few optimization passes know how to deal with the constrained intrinsics is still a valid choice for some.
> 
> There is no hope to produce code as performant as in default mode, but with strict exception semantics. So the only way to make a compiler suitable for all users is to provide means to tune it.

Is there a disagreement here? I don't see it. It's true that after we finish teaching all the passes about strictfp (constrained or bundles) we still won't have code that is as optimized if strict exception semantics are selected. This is inevitable since, for example, we can't hoist an FP instruction out of a loop unless we can prove it never traps when traps matter.

> > WRT eliminating the constrained intrinsics completely, I thought that operand bundles could only be attached to function calls and not regular instructions? If I'm wrong, we still have a problem because there are so many uses of the regular FP instructions that we can't be safe-by-default and still use those instructions. We'd need to keep some kind of the constrained intrinsics (or new intrinsics) that give us replacements for the regular FP instructions.
> 
> There is no intention to remove constrained intrinsics (like `experimental.constrained.fadd`), because operand bundles could only be attached to function calls. The constrained intrinsics however need to change their shape in future to use operand bundles as well.

Agreed, we'll need to keep some kind of intrinsics to replace the FP instructions. Since we're going to be replacing the metadata with bundles it might make sense to switch the names of the intrinsics while we're at it. That would also allow for parallel implementations during a switchover.

Changing to use bundles will be more flexible and can include things like the denormal mode. More flexibility sounds like a good change. It's a shame we'll lose "safe-by-default" for libm-style intrinsics, but I don't have an alternative.
 
> > The problem I've run into with the IR Verifier is that the strictfp mode of the IR Builder needs to be turned on in all cases needed, but currently doesn't because the IR Builder's constructor doesn't know to construct itself in strictfp mode. Now, many instantiations of the IR Builder occur with the constructor able to chase down the function definition and find out if strictfp mode is required. Many, hundreds, don't have that luxury and thus each has to be touched and corrected.
> 
> If IRBuilder does not set strictfp when it should, it usually ends up with assertion violation or verification errors. I guess in this case IR is created by some tool, probably MLIR-based, not by clang? Maybe this is a use case that we do not support enough well now and it would be helpful to investigate it.

No, there are hundreds of instantiations of the IRBuilder strewn all throughout llvm that don't properly set the strictfp mode when they should. Optimization passes that create new instructions (perhaps to replace one or more old instructions) use IRBuilder instances to create those instructions.

All IRBuilder instances need to be corrected before we can push any IR Verifier code that ensures correct use of the strictfp function and call attribute. Plus we still have tests that fail with the not-pushed-yet IR Verifier checks for strictfp, including one clang test. So people think that Verifier code is in the tree when it isn't yet.

I do have a ticket open to change the IRBuilder to check the function definition and set the strictfp mode automatically. But I pooched the branch on my end so I'll need to open a new ticket, hopefully this year. That still leaves hundreds of cases that need to be manually corrected.

spavloff committed:
Update tests in Transforms/EarlyCSE

spavloff committed:
Small changes: docs, clang-format, revievers' notes

- Fix Doxygen error,
- Fix clang-format error,
- remove unused function declaration,
- remove setting MD_fpmath, it is made by copyMetadata.

spavloff commented:
> > Right, we would need to introduce new llvm.fadd etc. to carry bundles. If there are no bundles these could fold back to the regular instruction
> 
> Good. The new intrinsics will be unknown to the optimization passes and thus be "safe by default".

Optimization passes can perform transformations based on the instruction properties rather than particular instructions. In such case constrained intrinsics can be involved in optimizations.

> > What do you mean by "safe-by-default"? Implementation must be correct, and produce the code that executes in accordance to standards and documentation. Any deviation from them is an error and should be fixed. Does it mean that such implementation would be "safe"?
> 
> When I say "safe-by-default" I'm referring to the case where the optimizers don't know about the intrinsics so the optimizers get very conservative and don't optimize.

What about users who need performance? For example, running code using non-default rounding mode should demonstrate almost the same performance as in the case of default mode, if the rounding mode is changed outside this code. The only way to represent such code now is using constrained functions. If they would not be optimized, performance would be poor.

> The sticky point is the rounding mode. Running with a different rounding mode ideally wouldn't affect performance much at all, but it currently does. But fixing that requires fixing all the optimization passes.

Exactly. To fix that the constrained intrinsics should be transformed in optimizations more actively.

> I do have a ticket open to change the IRBuilder to check the function definition and set the strictfp mode automatically. But I pooched the branch on my end so I'll need to open a new ticket, hopefully this year. That still leaves hundreds of cases that need to be manually corrected.

The attribute `strictfp` requires refinement, which is outside of the proposed changes and needs to be discussed separately. In particular a function that changes FP environmet, but then restores it should not require `strictfp` environment. It means `strictfp` should not be assigned automatically.

jcranmer-intel commented:
> We will need someone to go through and audit the libm-style intrinsics and make sure all optimizations of them are safe before we can switch over. Those are "unsafe by default" since the optimizers know about them already.

I've already been looking at all of the optimizations for fast-math flags as part of my work there.

Actually, most of the optimizations these days seem to be using the libcalls rather than the intrinsics (yay excessive duplication), so I don't think there's too much work with all of the libm intrinsics for optimizations.

kpneal commented:
> > I do have a ticket open to change the IRBuilder to check the function definition and set the strictfp mode automatically. But I pooched the branch on my end so I'll need to open a new ticket, hopefully this year. That still leaves hundreds of cases that need to be manually corrected.
> 
> The attribute `strictfp` requires refinement, which is outside of the proposed changes and needs to be discussed separately. In particular a function that changes FP environmet, but then restores it should not require `strictfp` environment. It means `strictfp` should not be assigned automatically.

Without the strictfp attribute the inliner will be allowed to inline that function that changes the FP environment into a function that is non-strictfp. Since we don't have FP barriers we'll then have optimizations moving instructions into and out of the inlined body. This can result in incorrect trap behavior, incorrect rounding happening, whatever, depending on how the FP environment is modified. So the strictfp attribute is required, period. What we need to do is reword the description in the LangRef to make this more clear.

spavloff commented:
PR https://github.com/llvm/llvm-project/pull/118253 demonstrates changes required to upgrade an intrinsic (`trunc` in this case) to using FP operand bundles.

spavloff commented:
### Merge activity

* **Dec 2, 1:35 AM EST**: The merge label 'FP Bundles' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=llvm&repo=llvm-project) once it meets the requirements.

kpneal commented on llvm/lib/IR/IntrinsicInst.cpp:
```diff
@@ -273,29 +306,6 @@ void InstrProfCallsite::setCallee(Value *Callee) {
-    return std::nullopt;
-  return convertStrToExceptionBehavior(cast<MDString>(MD)->getString());
-}
-
```
I thought the plan was to have concurrent implementations until we can throw the switch. That way we won't have releases that have half-broken constrained intrinsic support but half-implemented operand bundles.

spavloff committed:
Small fixes

- Add method `hasFloatingPointBundles` to simplify checks,
- Update comments in `TailRecursionElimination.cpp`,
- Updated messages in Verifier,
- Add Verifier tests.
- Rebame remained cases of 'fpe.round' for 'fpe.control'

spavloff committed:
Calculate MemoryEffect

spavloff committed:
Don't set strictfp on irrelevant calls

spavloff committed:
Revert "Don't set strictfp on irrelevant calls"

This change was discussed in
https://github.com/llvm/llvm-project/pull/122735. As it was decided,
current usage of strictfp on call sites remains intact.

spavloff committed:
Change documentation

spavloff committed:
Do not use strictfp call attribute as indicator of FPE awareness

spavloff committed:
Rename fpe.* to fp.*
